consensus/core/src/dag_state.rs

```diff
@@ -295,6 +379,31 @@ impl DagState {
         Ok(blocks)
     }
 
+    pub(crate) fn contains_block_at_slot(&self, slot: Slot) -> bool {
+        // Always return true for genesis slots.
+        if slot.round == GENESIS_ROUND {
+            return true;
+        }
+
+        // If the slot we are looking for is within the cached rounds set, then it's enough to only
+        // look in cached_refs and not in store. Otherwise we'll need to go to store
+        if let Some(lowest_round_ref) = self.cached_refs[slot.authority].first() {
```

Comment:
Relying on earliest ref is subtle. We have to make sure that if a (byzantine) block with round < authority's earliest ref round is accepted, DagState has to cache it at least temporarily, violating the invariant. Alternatively, we can check if the slot is above last commit round, and only check memory if it is. We can even use authority last commit round - CACHED_ROUND, but it will be harder to test for now.

Reply:
I see the case yes, but even under those circumstances we will still manage to get a correct answer as we'll fall back to store. Are we concerned here on the subtlety or the performance?

Reply:
Thanks for explaining the above @mwtian , I know see the issue and indeed how subtle this is, you are right. Let me refactor this as per the recommendation. I'll add some comments on the `recent_` structs on the data availability semantics because I think it would be quite easy to miss similar cases